### PR TITLE
WasmFS JS API: Implement utime

### PR DESCRIPTION
**Pull request by jameshu15869**
This PR implements `utime` for the JS API. I didn't find any tests for `utime` in the test directory, so added tests to `test_fs_js_api`. 

I also attempted to add support for `atime`, `mtime`, and `ctime` nanoseconds in WasmFS. I am still a little new to C++ programming, so please let me know about any improvements I need to make to the implementation. One thing I wasn't sure about was how to get the nanoseconds - is using `#include <time.h>` the best way?

### CODE REVIEW

File: test/fs/test_fs_js_api.c [hunk 1]
```diff
@@ -111,6 +111,36 @@ int main() {
         assert(ex.name === 'ErrnoError' && ex.errno == 8 /* EBADF */);
     );
 
+    /********** test FS.utime() **********/
```
Comment (sbc100) [inline test/fs/test_fs_js_api.c:114]: I wonder if its worth splitting this function up.  (That way the ones that are JS-only can also be written using EM_JS instead of EM_ASM)

Comment (sbc100) [reply test/fs/test_fs_js_api.c:114]: (Can be a followup PR for that)

File: src/library_wasmfs.js [hunk 2]
```diff
@@ -285,6 +285,12 @@ FS.createPreloadedFile = FS_createPreloadedFile;
     // TODO: truncate
     // TODO: ftruncate
     // TODO: utime
+    utime: (path, atime, mtime) => {
+      return FS.handleError(withStackSave(() => {
+        var pathBuffer = stringToUTF8OnStack(path);
+        return __wasmfs_utime(pathBuffer, atime, mtime, 0);
```
Comment (kripken) [inline src/library_wasmfs.js:291]: As this is a function for our internal use, I don't think we need to add the last parameter which is always 0 in our usage. (If we need it later, we can add it later - it's ok to change internal APIs.)

File: test/fs/test_fs_js_api.c (hunk 1 above)
Comment (jameshu15869) [reply test/fs/test_fs_js_api.c:114]: Just to clarify, do you mean splitting up by moving each test section into an `EM_JS` function? Or should the code be split some other way?

File: system/lib/wasmfs/file.h [hunk 3]
```diff
@@ -97,6 +98,10 @@ class File : public std::enable_shared_from_this<File> {
   File(FileKind kind, mode_t mode, backend_t backend)
     : kind(kind), mode(mode), backend(backend) {
     atime = mtime = ctime = time(NULL);
+
+    struct timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+    atime_nsec = mtime_nsec = ctime_nsec = ts.tv_nsec;
```
Comment (kripken) [inline system/lib/wasmfs/file.h:104]: It is more comprehensive to support nanoseconds as well, and I think this the right way to do it (but we could also get the non-nanosecond times from this as well, to save the previous `time(..)` call).

However, I think we need to consider if we want to add ns. Technically a POSIX-compliant FS doesn't need that (I don't think POSIX guarantees a particular precision). And on the Web the timing info we get from `Date.now()` etc. may have limited resolution anyhow.

And, adding ns support has a cost in terms of code size. So overall I lean towards not adding it, and just documenting that decision.

@tlively what do you think?

File: test/fs/test_fs_js_api.c (hunk 1 above)
Comment (sbc100) [reply test/fs/test_fs_js_api.c:114]: Yup thats what I meant.    Each section could be its own function (either a regular C function or an EM_JS function)

File: system/lib/wasmfs/file.h (hunk 3 above)
Comment (tlively) [reply system/lib/wasmfs/file.h:104]: +1 to removing the previous `time` call.

I would lean toward keeping this. I don't think it's too much of a stretch to imagine an Emscripten application that would want at least millisecond precision in the time stamps. For example, anything doing something similar to `make` to figure out when file changes mean something needs to be recomputed would need more than second precision.

I also expect that the code size increase would be fairly small.

File: system/lib/wasmfs/file.h [hunk 4]
```diff
@@ -315,11 +323,17 @@ class File::Handle {
...
   void setCTime(time_t time) { file->ctime = time; }
+  void setCTimeNs(long time) { file->ctime_nsec = time; }
   time_t getMTime() { return file->mtime; }
+  long getMTimeNs() { return file->mtime_nsec; }
   void setMTime(time_t time) { file->mtime = time; }
+  void setMTimeNs(long time) { file->mtime_nsec = time; }
   time_t getATime() { return file->atime; }
+  long getATimeNs() { return file->atime_nsec; }
   void setATime(time_t time) { file->atime = time; }
+  void setATimeNs(long time) { file->atime_nsec = time; }
```
Comment (tlively) [inline system/lib/wasmfs/file.h:336]: Maybe we can combine these getters and setters? I imagine that usually the seconds and nanoseconds will always be set or retrieved at the same time.

Comment (jameshu15869) [reply system/lib/wasmfs/file.h:336]: I think I was worried about compatibility with any code that used the old getters/setters by getting just the seconds with `time(NULL)`. Would it be worth switching all of them to use a combined getter/setter? If so, would it be better to pass in the seconds and nanoseconds together as a pointer to a struct or two separate `time_t` and `long` values respectively?

Comment (tlively) [reply system/lib/wasmfs/file.h]: Yes, switching all the use sites would be fine. There shouldn't be many of them, and this is private API so this wouldn't be a breaking change. I would implement it by passing the `timeval` struct by value (rather than by pointer) since it is so small.

File: test/fs/test_fs_js_api.c [hunk 5]
```diff
@@ -200,13 +224,28 @@ int main() {
...
+}
+
+int main() {
+    test_fs_open();
+    test_fs_rename();
+    test_fs_read();
+    test_fs_utime();
+    test_fs_rmdir();
+    test_fs_close();
+    test_fs_mknod();
```
Comment (sbc100) [inline test/fs/test_fs_js_api.c:246]: Nice, I think this makes the test much easier to read.

**Comment by jameshu15869** [comment]
Currently my implementation has failed `test_metadce_files_wasmfs` and `test_metadce_cxx_wasmfs`. I was able to reproduce the issue locally, but am not sure of the cause. Could it be that using `#include <time.h>` in `files.h` and `files.cpp` is causing the error?

File: system/lib/wasmfs/file.cpp [hunk 6]
```diff
@@ -204,7 +211,9 @@ int Directory::Handle::removeChild(const std::string& name) {
     entry->second.file->locked().setParent(nullptr);
     dcache.erase(entry);
   }
-  setMTime(time(NULL));
+  struct timespec ts;
+  clock_gettime(CLOCK_REALTIME, &ts);
+  setMTime(ts);
```
Comment (kripken) [inline system/lib/wasmfs/file.cpp:216]: It would be good to avoid this repeated code pattern somehow. If we did that then we would have a single place to change if we decided to get time information in another way in the future.

One option might be an `updateMTime` method which would mean "update to the current time". There might be a better name though?

Comment (kripken) [reply system/lib/wasmfs/file.cpp]: For symmetry it might be nice to have it for them all, even if mtime is the most-repeating pattern.

Comment (kripken) [reply system/lib/wasmfs/file.cpp]: Another option might be "updateTime" without mentioning mtime at all. Basically "this file was modified, update whatever time changes need to be done".

Comment (jameshu15869) [reply system/lib/wasmfs/file.cpp]: I thinking making one "update to current time" for each time might be a good option here, since there were a few other spots where only `atime` and `ctime` also had to be set to the current time. Would that be fine for now?

File: src/library_wasmfs.js [hunk 7]
```diff
@@ -286,6 +286,12 @@ FS.createPreloadedFile = FS_createPreloadedFile;
       return FS.handleError(__wasmfs_ftruncate(fd, {{{ splitI64('len') }}}));
     },
     // TODO: utime
+    utime: (path, atime, mtime) => {
+      return FS.handleError(withStackSave(() => {
```
Comment (sbc100) [inline src/library_wasmfs.js:290]: You can remove the `return` here and the curly brace before it.  Similar to #19614

File: test/fs/test_fs_js_api.c [hunk 8]
```diff
@@ -4,6 +4,27 @@
 #include <assert.h>
 #include <fcntl.h>
 
+EM_JS(void, test_fs_open, (), {
```
Comment (kripken) [inline test/fs/test_fs_js_api.c:7]: Maybe make a separate PR for the test refactorings of existing tests?

File: test/other/metadce/test_metadce_files_wasmfs.size [hunk 9]
```diff
@@ -1 +1 @@
-51960
+52403
```
Comment (kripken) [inline test/other/metadce/test_metadce_files_wasmfs.size:1]: This is the cost of supporting nanoseconds, I suppose?

I am not sure it's worth it, personally. 500 bytes isn't a lot, but the benefit is unclear to me... but if you feel strongly @tlively I don't object.

Regardless, if we decide we want nanoseconds let's do it in its own PR?

Comment (tlively) [reply test/other/metadce/test_metadce_files_wasmfs.size:1]: Yes, I think having better than second precision is worth 500 bytes. I'm surprised it costs that much, though.

File: src/library_wasmfs.js [hunk 10]
```diff
@@ -265,13 +265,19 @@ FS.createPreloadedFile = FS_createPreloadedFile;
     // TDOO: chown
     // TODO: lchown
     // TODO: fchown
+    // TODO: truncate
+    // TODO: ftruncate
```
Comment (tlively) [inline src/library_wasmfs.js:269]: Bad merge resolution?

**Review by tlively — APPROVED** [review]
LGTM % stale TODOs.

**Comment by kripken** [comment]
Is this ready to land? (it's still marked "draft")

**Comment by jameshu15869** [comment]
Oops, I forgot to set it as "ready" but the branch is ready to land.
